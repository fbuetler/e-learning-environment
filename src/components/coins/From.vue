<template>
  <div>
    <slot name="animation" :animationSteps="animationSteps" />
    <div>Welche Summe wird hier dargestellt?</div>
    <div id="coins" class="flex-item flex-center flex-col flex-flex card">
      <slot>
        <div class="flex-item flex-center flex-row">
          <div
            v-for="(amount, i) in generatedItems"
            :key="`item-${i}`"
            class="flex-item flex-center flex-col"
          >
            <div
              v-for="j in amount"
              :key="`amount-${j}`"
              :class="items(coinType)[i].class"
            >
              <img :src="require(`@/assets/${items(coinType)[i].img}`)" />
            </div>
          </div>
        </div>
      </slot>
    </div>
    <hr />
    <div>LÃ¶sung:</div>
    <div id="answer">
      <input
        id="answer-input"
        v-model.number="number"
        class="card big-text"
        size="5"
        type="number"
      />
    </div>
  </div>
</template>

<script lang="ts">
import { Component, Prop, Mixins } from "vue-property-decorator";
import GameMixin, { GameInterface } from "@/components/GameMixins.vue";
import CoinsMixin, { coinType } from "@/components/coins/CoinsMixin.vue";

@Component<From>({})
export default class From extends Mixins(GameMixin, CoinsMixin)
  implements GameInterface {
  @Prop({ required: true })
  args!: { coinType: coinType };

  coinType = this.args.coinType;

  number: number = null;
  solution: number = null;
  generatedItems: Array<number> = null;
  animationSteps: Array<string> = null;

  isStarted(): boolean {
    return this.generatedItems === null;
  }

  restartGame() {
    this.number = null;
    this.generatedItems = this.generateItems(this.coinType);
    this.solution = this.sumItems(this.coinType, this.generatedItems);
    this.animationSteps = this.getAnimationSteps();
  }

  isCorrect(): boolean {
    return this.number === this.solution;
  }

  getAnimationSteps(): Array<string> {
    return [
      `answer-input:${Math.ceil(Math.random() * this.limit)}`,
      "button-menu-check",
      `answer-input:${this.solution}`,
      "button-menu-check",
      "button-menu-next",
    ];
  }
}
</script>
