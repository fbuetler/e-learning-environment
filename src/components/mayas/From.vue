<template>
  <div>
    <slot name="animation" :animationSteps="animationSteps" />
    <div>Welche Zahl wird hier dargestellt?</div>
    <div class="flex-item flex-center flex-col flex-flex card">
      <div>
        <div class="flex-item flex-center flex-row">
          <div v-for="index in generatedItems[nut]" :key="index" class="nut">
            <img :src="require('@/assets/mayas/nut.png')" />
          </div>
        </div>
        <div class="flex-item flex-center flex-col">
          <div
            v-for="index in generatedItems[stick]"
            :key="index"
            class="stick"
          >
            <img :src="require('@/assets/mayas/stick.png')" />
          </div>
        </div>
      </div>
    </div>
    <hr />
    <div>LÃ¶sung:</div>
    <input
      id="answer-input"
      v-model.number="number"
      class="card big-text"
      size="5"
      type="number"
    />
  </div>
</template>

<script lang="ts">
import { Component, Mixins } from "vue-property-decorator";
import GameMixin, { GameInterface } from "@/components/GameMixins.vue";
import MayasMixin from "@/components/mayas/MayasMixin.vue";

@Component<From>({})
export default class From extends Mixins(GameMixin, MayasMixin)
  implements GameInterface {
  number: number = null;
  solution: number = null;
  generatedItems: Array<number> = null; // unfortunately Maps and Sets are not reactive in vue 2
  animationSteps: Array<string> = null;

  isStarted(): boolean {
    return this.generatedItems === null;
  }

  restartGame() {
    this.generatedItems = this.generateItems();
    this.solution = this.sumItems(this.generatedItems);
    this.animationSteps = this.getAnimationSteps();
    this.number = null;
  }

  isCorrect(): boolean {
    return this.number === this.solution;
  }

  getAnimationSteps(): Array<string> {
    return [
      `answer-input:${Math.ceil(Math.random() * this.limit)}`,
      "button-menu-check",
      `answer-input:${this.solution}`,
      "button-menu-check",
      "button-menu-next",
    ];
  }
}
</script>
