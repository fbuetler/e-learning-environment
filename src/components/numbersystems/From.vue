<template>
  <div>
    <slot name="animation" :animationSteps="animationSteps" />
    <div>Welche Summe wird hier dargestellt?</div>
    <ItemGroup :items="generatedItems" :itemConfig="items(type)" />
    <hr />
    <div>LÃ¶sung:</div>
    <div id="answer">
      <input
        id="answer-input"
        v-model.number="number"
        class="card big-text"
        size="5"
        type="number"
      />
    </div>
  </div>
</template>

<script lang="ts">
import { Component, Prop, Mixins } from "vue-property-decorator";
import GameMixin, { GameInterface } from "@/components/GameMixins.vue";
import NumbersystemsMixin, {
  numbersystemType,
} from "@/components/numbersystems/NumbersystemsMixin.vue";
import ItemGroup from "@/components/numbersystems/ItemGroup.vue";

@Component<From>({
  components: {
    ItemGroup,
  },
})
export default class From extends Mixins(GameMixin, NumbersystemsMixin)
  implements GameInterface {
  @Prop({ required: true })
  args!: { numbersystemType: numbersystemType };
  type = this.args.numbersystemType;

  number: number = null;
  solution: number = null;
  generatedItems: Array<number> = null;
  animationSteps: Array<string> = null;

  isInitialized(): boolean {
    return this.generatedItems !== null;
  }

  start() {
    this.number = null;
    this.generatedItems = this.generateItems(this.type);
    this.solution = this.sumItems(this.type, this.generatedItems);
    this.animationSteps = this.getAnimationSteps();
  }

  isCorrect(): boolean {
    return this.number === this.solution;
  }

  getAnimationSteps(): Array<string> {
    return [
      `answer-input:${Math.ceil(Math.random() * this.limit(this.type))}`,
      "button-menu-check",
      `answer-input:${this.solution}`,
      "button-menu-check",
      "button-menu-next",
    ];
  }
}
</script>
