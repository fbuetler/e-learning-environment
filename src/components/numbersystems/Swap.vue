<template>
  <div>
    <slot name="animation" :animationSteps="animationSteps" />
    <div>Stelle die gleiche Summe mit weniger Elementen dar</div>
    <ItemGroup :items="generatedItems" :itemConfig="items(type)" />
    <hr />
    <div>LÃ¶sung:</div>
    <ItemDropzone
      :items="selectedItems"
      :itemConfig="items(type)"
      @dropped="addItem()"
    />
    <div
      class="interaction-container flex-item flex-row flex-center flex-stretch"
    >
      <ItemSelection
        :selected="selected"
        :items="items(type)"
        @selected="selected = $event"
      />
      <Undo @undo-operation="undo()" />
    </div>
  </div>
</template>

<script lang="ts">
import { Vue, Component, Prop, Mixins } from "vue-property-decorator";
import GameMixin, { GameInterface } from "@/components/GameMixins.vue";
import NumbersystemsMixin, {
  numbersystemType,
} from "@/components/numbersystems/NumbersystemsMixin.vue";
import Undo from "@/components/Undo.vue";
import Difficulty from "@/components/Difficulty.vue";
import ItemSelection from "@/components/ItemSelection.vue";
import ItemDropzone from "@/components/numbersystems/ItemDropzone.vue";
import ItemGroup from "@/components/numbersystems/ItemGroup.vue";

@Component<Swap>({
  components: {
    Undo,
    Difficulty,
    ItemSelection,
    ItemDropzone,
    ItemGroup,
  },
})
export default class Swap extends Mixins(GameMixin, NumbersystemsMixin)
  implements GameInterface {
  @Prop({ required: true })
  args!: { numbersystemType: numbersystemType };
  type = this.args.numbersystemType;

  selected: number = null;
  selectedItems: Array<number> = null;
  generatedItems: Array<number> = null;
  animationSteps: Array<string> = null;

  currentDifficultyLevel = 1;
  difficultyLevels = 2;

  isInitialized(): boolean {
    return this.generatedItems !== null;
  }

  start() {
    this.selected = null;
    this.selectedItems = new Array<number>(this.items(this.type).length).fill(
      0
    );
    do {
      this.generatedItems = this.generateItems(this.type);
    } while (
      this.sumItems(this.type, this.generatedItems) >= this.limit(this.type) ||
      this.sum(this.generatedItems) ===
        this.sum(
          this.calcMinimalAmount(
            this.type,
            this.sumItems(this.type, this.generatedItems)
          )
        )
    );
    this.animationSteps = this.getAnimationSteps();
  }

  isCorrect(): boolean {
    return (
      this.sumItems(this.type, this.selectedItems) ===
        this.sumItems(this.type, this.generatedItems) &&
      this.sum(this.selectedItems) < this.sum(this.generatedItems)
    );
  }

  sum(arr: number[]): number {
    return arr.reduce((sum, el) => (sum += el));
  }

  addItem() {
    if (this.selected === null) {
      return;
    }
    Vue.set(
      this.selectedItems,
      this.selected,
      this.selectedItems[this.selected] + 1
    );
    this.selected = null;
  }

  undo() {
    for (let i = 0; i < this.selectedItems.length; i++) {
      Vue.set(this.selectedItems, i, 0);
    }
  }

  getAnimationSteps(): Array<string> {
    const correctSum = this.sumItems(this.type, this.generatedItems);
    const wrongSum = Math.ceil(Math.random() * this.limit(this.type));

    return this.mapNumberToActions(wrongSum, this.type)
      .concat(["button-menu-check", "undo"])
      .concat(this.mapNumberToActions(correctSum, this.type))
      .concat(["button-menu-check", "button-menu-next"]);
  }
}
</script>

<style scoped></style>
