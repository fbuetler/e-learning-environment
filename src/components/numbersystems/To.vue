<template>
  <div @dragend.prevent="selected = null">
    <slot name="animation" :animationSteps="animationSteps" />
    <Difficulty
      v-if="displayDifficulty"
      :selected="currentDifficultyLevel"
      :difficultyLevels="difficultyLevels"
      @difficulty-selected="currentDifficultyLevel = $event"
    />
    <div class="number">
      Stell die folgende Summe
      <b v-if="currentDifficultyLevel === 2">mit möglichst wenig Münzen</b> dar:
    </div>
    <div id="number" class="big-text">{{ number }}</div>
    <div
      id="dropzone"
      class="flex-item flex-center flex-col flex-flex dropzone"
      @click="addItem()"
      @dragover.prevent
      @dragend.prevent
      @drop.stop.prevent="addItem()"
    >
      <div v-if="nothingSelected">
        Platziere hier die Münzen
      </div>
      <div v-else>
        <slot>
          <div class="flex-item flex-center flex-row">
            <div
              v-for="(amount, i) in selectedItems"
              :key="`item-${i}`"
              class="flex-item flex-center flex-col"
            >
              <div
                v-for="j in amount"
                :key="`amount-${j}`"
                :class="items(type)[i].class"
              >
                <img :src="require(`@/assets/${items(type)[i].img}`)" />
              </div>
            </div>
          </div>
        </slot>
      </div>
    </div>
    <hr />
    <div
      class="interaction-container flex-item flex-row flex-center flex-stretch"
    >
      <ItemSelection
        :selected="selected"
        :items="items(type)"
        @selected="selected = $event"
      />
      <Undo @undo-operation="undo()" />
    </div>
  </div>
</template>

<script lang="ts">
import { Vue, Component, Prop, Mixins } from "vue-property-decorator";
import GameMixin, { GameInterface } from "@/components/GameMixins.vue";
import NumbersystemsMixin, {
  numbersystemType,
} from "@/components/numbersystems/NumbersystemsMixin.vue";
import ItemSelection from "@/components/ItemSelection.vue";
import Undo from "@/components/Undo.vue";
import Difficulty from "@/components/Difficulty.vue";

@Component<To>({
  components: {
    ItemSelection,
    Undo,
    Difficulty,
  },
})
export default class To extends Mixins(GameMixin, NumbersystemsMixin)
  implements GameInterface {
  @Prop({ required: true })
  args!: { numbersystemType: numbersystemType };
  type = this.args.numbersystemType;

  selected: number = null;
  number: number = null;
  selectedItems: Array<number> = null;
  animationSteps: Array<string> = null;

  // there are only difficulty levels for numbersystemType.DECIMAL:
  // represent number with minimal amount of items
  currentDifficultyLevel = 1;
  difficultyLevels = 2;

  isStarted(): boolean {
    return this.number === null;
  }

  start() {
    this.number = Math.ceil(this.randomNumber(this.limit(this.type)));
    this.selected = null;
    this.selectedItems = new Array<number>(this.items(this.type).length).fill(
      0
    );
    this.animationSteps = this.getAnimationSteps();
  }

  isCorrect(): boolean {
    if (
      this.type === numbersystemType.BINARY &&
      this.selectedItems.some((el) => el > 1)
    ) {
      return false;
    }
    const isCorrectSum =
      this.sumItems(this.type, this.selectedItems) === this.number;
    if (
      this.currentDifficultyLevel === 2 &&
      this.type === numbersystemType.DECIMAL &&
      isCorrectSum
    ) {
      return (
        JSON.stringify(this.selectedItems) ===
        JSON.stringify(this.calcMinimalAmount(this.type, this.number))
      );
    }
    return isCorrectSum;
  }

  addItem() {
    if (this.selected === null) {
      return;
    }
    Vue.set(
      this.selectedItems,
      this.selected,
      this.selectedItems[this.selected] + 1
    );
    this.selected = null;
  }

  undo() {
    for (let i = 0; i < this.selectedItems.length; i++) {
      Vue.set(this.selectedItems, i, 0);
    }
  }

  getAnimationSteps(): Array<string> {
    const correctNumber = this.number;
    const wrongNumber = Math.ceil(this.randomNumber(this.limit(this.type)));

    return this.mapNumberToActions(wrongNumber, this.type)
      .concat(["button-menu-check", "undo"])
      .concat(this.mapNumberToActions(correctNumber, this.type))
      .concat(["button-menu-check", "button-menu-next"]);
  }

  get nothingSelected(): boolean {
    return this.selectedItems.every((el) => el === 0);
  }

  get displayDifficulty(): boolean {
    return this.type === numbersystemType.DECIMAL;
  }
}
</script>

<style scoped>
.number {
  margin: 1em;
}
.dropzone {
  min-height: 5em;
}
</style>
