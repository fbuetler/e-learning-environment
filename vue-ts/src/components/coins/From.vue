<template>
  <div>
    <div>Welche Zahl wird hier dargestellt?</div>
    <div class="flex-item flex-center flex-col flex-flex card">
      <slot>
        <div class="flex-item flex-center flex-row">
          <div
            class="flex-item flex-center flex-col"
            v-for="(amount, i) in generatedItems"
            :key="`item-${i}`"
          >
            <div
              :class="items(coinType)[i].class"
              v-for="j in amount"
              :key="`amount-${j}`"
            >
              <img :src="require(`@/assets/${items(coinType)[i].img}`)" />
            </div>
          </div>
        </div>
      </slot>
    </div>
    <hr />
    <div>LÃ¶sung:</div>
    <input
      class="card big-text"
      size="5"
      v-model.number="number"
      type="number"
    />
  </div>
</template>

<script lang="ts">
import { Component, Prop, Mixins } from "vue-property-decorator";
import GameMixin, { GameInterface } from "@/components/GameMixins.vue";
import CoinsMixin, { coinType } from "@/components/coins/CoinsMixin.vue";

/*
  TODO
    - add tutorial animation
*/

@Component<From>({})
export default class From extends Mixins(GameMixin, CoinsMixin)
  implements GameInterface {
  @Prop({ required: true })
  args!: { coinType: coinType };

  coinType = this.args.coinType;
  number: number = null;
  generatedItems: Array<number> = null;

  isStarted(): boolean {
    return this.generatedItems === null;
  }

  restartGame() {
    this.generatedItems = this.generateItems(this.coinType);
  }

  isCorrect(): boolean {
    return this.number === this.sumItems(this.coinType, this.generatedItems);
  }
}
</script>
