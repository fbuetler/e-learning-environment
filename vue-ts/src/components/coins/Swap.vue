<template>
  <div>
    <slot name="animation" :animationSteps="animationSteps" />
    <div>Stelle die gleiche Summe mit weniger Münzen dar</div>
    <div id="coins" class="flex-item flex-center flex-col flex-flex card">
      <slot>
        <div class="flex-item flex-center flex-row">
          <div
            v-for="(amount, i) in generatedItems"
            :key="`item-${i}`"
            class="flex-item flex-center flex-col"
          >
            <div
              v-for="j in amount"
              :key="`amount-${j}`"
              :class="items(type)[i].class"
            >
              <img :src="require(`@/assets/${items(type)[i].img}`)" />
            </div>
          </div>
        </div>
      </slot>
    </div>
    <hr />
    <div>Lösung:</div>
    <div
      id="dropzone"
      class="flex-item flex-center flex-col flex-flex dropzone"
      @click="addItem()"
      @dragover.prevent
      @dragend.prevent
      @drop.stop.prevent="addItem()"
    >
      <div v-if="nothingSelected">
        Platziere hier die Münzen
      </div>
      <div v-else>
        <slot>
          <div class="flex-item flex-center flex-row">
            <div
              v-for="(amount, i) in displaySelectedItems"
              :key="`item-${i}`"
              class="flex-item flex-center flex-col"
            >
              <div
                v-for="j in amount"
                :key="`amount-${j}`"
                :class="items(type).reverse()[i].class"
              >
                <img
                  :src="require(`@/assets/${items(type).reverse()[i].img}`)"
                />
              </div>
            </div>
          </div>
        </slot>
      </div>
    </div>
    <div
      class="interaction-container flex-item flex-row flex-center flex-stretch"
    >
      <ItemSelection
        :selected="selected"
        :items="items(type).reverse()"
        @selected="selected = $event"
      />
      <Undo @undo-operation="undo()" />
    </div>
  </div>
</template>

<script lang="ts">
import { Vue, Component, Mixins } from "vue-property-decorator";
import GameMixin, { GameInterface } from "@/components/GameMixins.vue";
import CoinsMixin, { coinType } from "@/components/coins/CoinsMixin.vue";
import ItemSelection from "@/components/ItemSelection.vue";
import Undo from "@/components/Undo.vue";
import Difficulty from "@/components/Difficulty.vue";

// TODO: (optional) less coins or minimal number of coins? -> both: difficulties

@Component<Swap>({
  components: {
    ItemSelection,
    Undo,
    Difficulty,
  },
})
export default class Swap extends Mixins(GameMixin, CoinsMixin)
  implements GameInterface {
  type = coinType.NORMAL;

  selected: number = null;
  selectedItems: Array<number> = null;
  generatedItems: Array<number> = null;
  animationSteps: Array<string> = null;

  currentDifficultyLevel = 1;
  difficultyLevels = 2;

  isStarted(): boolean {
    return this.generatedItems === null;
  }

  restartGame() {
    this.selected = null;
    this.selectedItems = new Array<number>(this.items(this.type).length).fill(
      0
    );
    do {
      this.generatedItems = this.generateItems(this.type);
    } while (
      this.countCoins(this.generatedItems) ===
      this.countCoins(
        this.calcMinimalAmount(
          this.sumItems(this.type, this.generatedItems),
          this.type
        )
      )
    );
    this.animationSteps = this.getAnimationSteps();
  }

  isCorrect(): boolean {
    return (
      this.sumItems(this.type, this.selectedItems) ===
        this.sumItems(this.type, this.generatedItems) &&
      this.countCoins(this.selectedItems) < this.countCoins(this.generatedItems)
    );
  }

  countCoins(arr: number[]): number {
    return arr.reduce((sum, el) => (sum += el));
  }

  addItem() {
    if (this.selected === null) {
      return;
    }
    Vue.set(
      this.selectedItems,
      this.selected,
      this.selectedItems[this.selected] + 1
    );
    this.selected = null;
  }

  undo() {
    for (let i = 0; i < this.selectedItems.length; i++) {
      Vue.set(this.selectedItems, i, 0);
    }
  }

  getAnimationSteps(): Array<string> {
    const correctSum = this.sumItems(this.type, this.generatedItems);
    const wrongSum = Math.ceil(Math.random() * this.limit);

    return this.mapNumberToActions(wrongSum, this.type)
      .concat(["button-menu-check", "undo"])
      .concat(this.mapNumberToActions(correctSum, this.type))
      .concat(["button-menu-check", "button-menu-next"]);
  }

  get nothingSelected(): boolean {
    return this.selectedItems.every((el) => el === 0);
  }

  get displaySelectedItems(): number[] {
    return this.selectedItems.slice().reverse();
  }
}
</script>

<style scoped>
.dropzone {
  min-height: 5em;
}
</style>
