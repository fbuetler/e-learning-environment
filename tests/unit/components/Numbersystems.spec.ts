import { shallowMount, config } from "@vue/test-utils";
import To from "@/components/numbersystems/To.vue";
import From from "@/components/numbersystems/From.vue";
import Swap from "@/components/numbersystems/Swap.vue";

config.showDeprecationWarnings = false;

enum numbersystemsType {
  DECIMAL,
  BINARY,
  MAYA,
}

const mockNumber = 27;
const mockDecimalSolution = [0, 1, 0, 1, 1, 0]; // 20 + 5 + 2= 27
const mockRandomNumber = jest.fn((x) => mockNumber);
const mockItems = [0, 0, 1, 1, 5, 2]; //  + 1*10 + 1*5 + 5*2 + 2*1 = 27
const mockGenerateItems = jest.fn((x) => mockItems);

describe("To.vue", () => {
  let wrapper;
  beforeEach(() => {
    wrapper = shallowMount(To, {
      propsData: {
        args: {
          numbersystemType: numbersystemsType.DECIMAL,
        },
      },
      methods: {
        randomNumber: mockRandomNumber,
      },
    });
  });

  it("is a Vue instance", () => {
    expect(wrapper.vm).toBeTruthy();
  });

  it("renders correctly", () => {
    expect(wrapper.element).toMatchSnapshot();
  });

  it("initial conditions hold", () => {
    expect(wrapper.vm["number"]).toBe(mockNumber);
    expect(wrapper.vm["selected"]).toBeNull();
    expect(wrapper.vm["selectedItems"].every((el) => el === 0)).toBeTruthy();
  });

  it("next task restores initial conditions", async () => {
    wrapper.setData({ selected: 0 });
    await wrapper.find("#dropzone").trigger("click");
    wrapper.vm.start();
    expect(wrapper.vm["number"]).toBe(mockNumber);
    expect(wrapper.vm["selected"]).toBeNull();
    expect(wrapper.vm["selectedItems"].every((el) => el === 0)).toBeTruthy();
  });

  it("undo restores initial conditions", async () => {
    wrapper.setData({ selected: 0 });
    await wrapper.find("#dropzone").trigger("click");
    wrapper.vm.undo();
    expect(wrapper.vm["number"]).toBe(mockNumber);
    expect(wrapper.vm["selected"]).toBeNull();
    expect(wrapper.vm["selectedItems"].every((el) => el === 0)).toBeTruthy();
  });

  it("correct result is accepted", async () => {
    for (let i = 0; i < mockDecimalSolution.length; i++) {
      for (let j = 0; j < mockDecimalSolution[i]; j++) {
        wrapper.setData({ selected: i });
        await wrapper.find("#dropzone").trigger("click");
      }
    }
    expect(wrapper.vm.isCorrect()).toBeTruthy();
  });

  it("incorrect result is rejected", async () => {
    wrapper.setData({ selected: 0 });
    await wrapper.find("#dropzone").trigger("click");
    expect(wrapper.vm.isCorrect()).toBeFalsy;
  });
});

describe("From.vue", () => {
  let wrapper;
  beforeEach(() => {
    wrapper = shallowMount(From, {
      propsData: {
        args: {
          numbersystemType: numbersystemsType.DECIMAL,
        },
      },
      methods: {
        generateItems: mockGenerateItems,
      },
    });
  });

  it("is a Vue instance", () => {
    expect(wrapper.vm).toBeTruthy();
  });

  it("renders correctly", () => {
    expect(wrapper.element).toMatchSnapshot();
  });

  it("initial conditions hold", () => {
    expect(wrapper.vm["number"]).toBeNull();
    expect(wrapper.vm["generatedItems"]).toEqual(mockItems);
    expect(wrapper.vm["solution"]).toBe(mockNumber);
  });

  it("next task restores initial conditions", () => {
    wrapper.setData({ number: 3 });
    wrapper.vm.start();
    expect(wrapper.vm["number"]).toBeNull();
    expect(wrapper.vm["generatedItems"]).toEqual(mockItems);
    expect(wrapper.vm["solution"]).toBe(mockNumber);
  });

  it("correct result is accepted", () => {
    wrapper.setData({ number: mockNumber });
    expect(wrapper.vm.isCorrect()).toBeTruthy;
  });

  it("incorrect result is rejected", () => {
    wrapper.setData({ number: 3 });
    expect(wrapper.vm.isCorrect()).toBeTruthy;
  });
});

describe("Swap.vue", () => {
  let wrapper;
  beforeEach(() => {
    wrapper = shallowMount(Swap, {
      propsData: {
        args: {
          numbersystemType: numbersystemsType.DECIMAL,
        },
      },
      methods: {
        generateItems: mockGenerateItems,
      },
    });
  });

  it("is a Vue instance", () => {
    expect(wrapper.vm).toBeTruthy();
  });

  it("renders correctly", () => {
    expect(wrapper.element).toMatchSnapshot();
  });

  it("initial conditions hold", () => {
    expect(wrapper.vm["selected"]).toBeNull();
    expect(wrapper.vm["selectedItems"].every((el) => el === 0)).toBeTruthy();
    expect(wrapper.vm["generatedItems"]).toEqual(mockItems);
  });

  it("next task restores initial conditions", async () => {
    wrapper.setData({ selected: 0 });
    await wrapper.find("#dropzone").trigger("click");
    wrapper.vm.start();
    expect(wrapper.vm["selected"]).toBeNull();
    expect(wrapper.vm["selectedItems"].every((el) => el === 0)).toBeTruthy();
    expect(wrapper.vm["generatedItems"]).toEqual(mockItems);
  });

  it("undo restores initial conditions", async () => {
    wrapper.setData({ selected: 0 });
    await wrapper.find("#dropzone").trigger("click");
    wrapper.vm.start();
    expect(wrapper.vm["selected"]).toBeNull();
    expect(wrapper.vm["selectedItems"].every((el) => el === 0)).toBeTruthy();
    expect(wrapper.vm["generatedItems"]).toEqual(mockItems);
  });

  it("correct result is accepted", async () => {
    for (let i = 0; i < mockDecimalSolution.length; i++) {
      for (let j = 0; j < mockDecimalSolution[i]; j++) {
        wrapper.setData({ selected: i });
        await wrapper.find("#dropzone").trigger("click");
      }
    }
    expect(wrapper.vm.isCorrect()).toBeTruthy();
  });

  it("incorrect result is rejected (correct sum, but not less items)", async () => {
    for (let i = 0; i < mockItems.length; i++) {
      for (let j = 0; j < mockItems[i]; j++) {
        wrapper.setData({ selected: i });
        await wrapper.find("#dropzone").trigger("click");
      }
    }
    expect(wrapper.vm.isCorrect()).toBeFalsy();
  });

  it("incorrect result is rejected (wrong sum)", async () => {
    wrapper.setData({ selected: 0 });
    await wrapper.find("#dropzone").trigger("click");
    expect(wrapper.vm.isCorrect()).toBeFalsy();
  });
});

describe("Addition.vue", () => {
  // TODO add tests
});
