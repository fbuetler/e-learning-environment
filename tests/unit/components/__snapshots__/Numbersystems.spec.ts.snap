// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`From.vue renders correctly 1`] = `
<div>
   
  <div>
    Welche Summe wird hier dargestellt?
  </div>
   
  <div
    class="flex-item flex-center flex-col flex-flex card"
    id="items"
  >
    <div
      class="flex-item flex-center flex-row"
    >
      <div
        class="flex-item flex-center flex-col"
      />
      <div
        class="flex-item flex-center flex-col"
      />
      <div
        class="flex-item flex-center flex-col"
      >
        <div
          class="responsive"
        >
          <img
            src=""
          />
        </div>
      </div>
      <div
        class="flex-item flex-center flex-col"
      >
        <div
          class="responsive"
        >
          <img
            src=""
          />
        </div>
      </div>
      <div
        class="flex-item flex-center flex-col"
      >
        <div
          class="responsive"
        >
          <img
            src=""
          />
        </div>
        <div
          class="responsive"
        >
          <img
            src=""
          />
        </div>
        <div
          class="responsive"
        >
          <img
            src=""
          />
        </div>
        <div
          class="responsive"
        >
          <img
            src=""
          />
        </div>
        <div
          class="responsive"
        >
          <img
            src=""
          />
        </div>
      </div>
      <div
        class="flex-item flex-center flex-col"
      >
        <div
          class="responsive"
        >
          <img
            src=""
          />
        </div>
        <div
          class="responsive"
        >
          <img
            src=""
          />
        </div>
      </div>
    </div>
  </div>
   
  <hr />
   
  <div>
    Lösung:
  </div>
   
  <div
    id="answer"
  >
    <input
      class="card big-text"
      id="answer-input"
      size="5"
      type="number"
    />
  </div>
</div>
`;

exports[`Swap.vue renders correctly 1`] = `
<div>
   
  <div>
    Stelle die gleiche Summe mit weniger Elementen dar
  </div>
   
  <div
    class="flex-item flex-center flex-col flex-flex card"
    id="items"
  >
    <div
      class="flex-item flex-center flex-row"
    >
      <div
        class="flex-item flex-center flex-col"
      />
      <div
        class="flex-item flex-center flex-col"
      />
      <div
        class="flex-item flex-center flex-col"
      >
        <div
          class="responsive"
        >
          <img
            src=""
          />
        </div>
      </div>
      <div
        class="flex-item flex-center flex-col"
      >
        <div
          class="responsive"
        >
          <img
            src=""
          />
        </div>
      </div>
      <div
        class="flex-item flex-center flex-col"
      >
        <div
          class="responsive"
        >
          <img
            src=""
          />
        </div>
        <div
          class="responsive"
        >
          <img
            src=""
          />
        </div>
        <div
          class="responsive"
        >
          <img
            src=""
          />
        </div>
        <div
          class="responsive"
        >
          <img
            src=""
          />
        </div>
        <div
          class="responsive"
        >
          <img
            src=""
          />
        </div>
      </div>
      <div
        class="flex-item flex-center flex-col"
      >
        <div
          class="responsive"
        >
          <img
            src=""
          />
        </div>
        <div
          class="responsive"
        >
          <img
            src=""
          />
        </div>
      </div>
    </div>
  </div>
   
  <hr />
   
  <div>
    Lösung:
  </div>
   
  <div
    class="flex-item flex-center flex-col flex-flex dropzone"
    id="dropzone"
  >
    <div>
      
    Platziere hier die Elemente
  
    </div>
  </div>
   
  <div
    class="interaction-container flex-item flex-row flex-center flex-stretch flex-wrap"
  >
    <div
      class="flex-item flex-wrap flex-center-horizontally flex-row"
      id="item-selection"
    >
      <div
        class="flex-item flex-wrap flex-center card clickable responsive"
        draggable="true"
        id="item-selection-6"
      >
        <img
          src=""
        />
      </div>
      <div
        class="flex-item flex-wrap flex-center card clickable responsive"
        draggable="true"
        id="item-selection-5"
      >
        <img
          src=""
        />
      </div>
      <div
        class="flex-item flex-wrap flex-center card clickable responsive"
        draggable="true"
        id="item-selection-4"
      >
        <img
          src=""
        />
      </div>
      <div
        class="flex-item flex-wrap flex-center card clickable responsive"
        draggable="true"
        id="item-selection-3"
      >
        <img
          src=""
        />
      </div>
      <div
        class="flex-item flex-wrap flex-center card clickable responsive"
        draggable="true"
        id="item-selection-2"
      >
        <img
          src=""
        />
      </div>
      <div
        class="flex-item flex-wrap flex-center card clickable responsive"
        draggable="true"
        id="item-selection-1"
      >
        <img
          src=""
        />
      </div>
    </div>
     
    <div
      class="flex-item flex-center card clickable responsive"
      id="undo"
      title="Neu starten!"
    >
      <img
        src=""
      />
    </div>
  </div>
</div>
`;

exports[`To.vue renders correctly 1`] = `
<div>
   
  <div
    class="flex-item flex-center flex-col"
  >
    <div>
      Ändere den Schwierigkeitsgrad:
    </div>
     
    <div
      class="flex-item flex-center flex-row"
    >
      <div
        class="card clickable responsive selected"
        id="difficulty-1"
        title="Schwierigkeitsgrad: leicht "
      >
        <img
          src=""
        />
      </div>
      <div
        class="card clickable responsive"
        id="difficulty-2"
        title="Schwierigkeitsgrad: mittelschwer "
      >
        <img
          src=""
        />
      </div>
    </div>
  </div>
   
  <div
    class="number"
  >
    
    Zahle mit
    
    <!---->
     
    <!---->
    
    Münzen die folgende Summe:
  
  </div>
   
  <div
    class="big-text"
    id="number"
  >
    27
  </div>
   
  <div
    class="flex-item flex-center flex-col flex-flex dropzone"
    id="dropzone"
  >
    <div>
      
    Platziere hier die Elemente
  
    </div>
  </div>
   
  <hr />
   
  <div
    class="interaction-container flex-item flex-row flex-center flex-stretch flex-wrap"
  >
    <div
      class="flex-item flex-wrap flex-center-horizontally flex-row"
      id="item-selection"
    >
      <div
        class="flex-item flex-wrap flex-center card clickable responsive"
        draggable="true"
        id="item-selection-6"
      >
        <img
          src=""
        />
      </div>
      <div
        class="flex-item flex-wrap flex-center card clickable responsive"
        draggable="true"
        id="item-selection-5"
      >
        <img
          src=""
        />
      </div>
      <div
        class="flex-item flex-wrap flex-center card clickable responsive"
        draggable="true"
        id="item-selection-4"
      >
        <img
          src=""
        />
      </div>
      <div
        class="flex-item flex-wrap flex-center card clickable responsive"
        draggable="true"
        id="item-selection-3"
      >
        <img
          src=""
        />
      </div>
      <div
        class="flex-item flex-wrap flex-center card clickable responsive"
        draggable="true"
        id="item-selection-2"
      >
        <img
          src=""
        />
      </div>
      <div
        class="flex-item flex-wrap flex-center card clickable responsive"
        draggable="true"
        id="item-selection-1"
      >
        <img
          src=""
        />
      </div>
    </div>
     
    <div
      class="flex-item flex-center card clickable responsive"
      id="undo"
      title="Neu starten!"
    >
      <img
        src=""
      />
    </div>
  </div>
</div>
`;
